pvlib/mlfm.py

```diff
@@ -0,0 +1,804 @@
+            'stack_i_ff',
+            'stack_r_sc',
+            'stack_i_sc']
+        color_map = [
+            'white',  # colour to bottom of graph
+            clr['temp_module'],
+            clr['v_oc'],
+            clr['r_oc'],
+            clr['v_ff'],
+            clr['i_v'],
+            clr['i_ff'],
+            clr['r_sc'],
+            clr['i_sc']]
+
+    if all([c in dstack.columns for c in stack4]):
+        # data order from bottom to top
+        ydata = [dnorm['pr_dc'] + (dstack['i_sc'] * (is_i_sc_self_ref)),
+                 dstack['temp_module_corr'] * (is_v_oc_temp_module_corr),
+                 dstack['v_oc'] - (
+                     dstack['temp_module_corr'] * (is_v_oc_temp_module_corr)),
+                 dstack['v_mp'],
+                 dstack['i_v'],
+                 dstack['i_mp'],
+                 dstack['i_sc'] * (not is_i_sc_self_ref)]
+        labels = [
+            'pr_dc',
+            'stack_t_mod',
+            'stack_v_oc',
+            'stack_v_mp',
+            '- - -',
+            'stack_i_mp',
+            'stack_i_sc']
+        color_map = [
+            'white',  # colour to bottom of graph
+            clr['temp_module'],
+            clr['v_oc'],
+            clr['v_mp'],
+            clr['i_v'],
+            clr['i_mp'],
+            clr['i_sc']]
+
+    # offset legend right, use ~1.2
+    bbox = 1.2
+
+    # select x axis usually date_time
+    xdata = dmeas.index.values
+    fig, ax1 = plt.subplots()
+
+    ax1.set_title(title)
+
+    # plot stack in order bottom to top,
+    # allowing self_ref and temp_module corrections
+    ax1.stackplot(xdata, *tuple(ydata), labels=labels, colors=color_map)
+
+    ax1.axhline(y=1/fill_factor, c='grey', lw=3)  # show initial 1/FF
+    ax1.axhline(y=1, c='grey', lw=3)  # show 100% line
+    ax1.set_ylabel('stacked mlfm losses')
+
+    # find number of x date values
+    x_ticks = dmeas.shape[0]
+    plt.xticks(np.arange(0, x_ticks), rotation=90)
+
+    # if (xaxis_labels > 0 and xaxis_labels < x_ticks):
+    if 0 < xaxis_labels < x_ticks:
+        xaxis_skip = np.floor(x_ticks / xaxis_labels)
+    else:
+        xaxis_skip = 2
+
+    #
+    xax2 = [''] * x_ticks
+    x_count = 0
+    while x_count < x_ticks:
+        if x_count % xaxis_skip == 0:
+            #
+            #  try to reformat any date indexes (not for matrices)
+            #
+            #  0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9
+            #   y y y y - m m - d d t h h : m m : s s --> yy-mm-dd hh'h'
+            #
+            try:
+                xax2[x_count] = xdata[x_count][2:13]+'h'
+            except IndexError:
+                xax2[x_count] = xdata[x_count]
+            except TypeError:  # xdata can't be subscripted
+                xax2[x_count] = xdata[0]
+
+        x_count += 1
+
+    ax1.set_xticklabels(xax2)
+    ax1.set_ylim(0.6, 1/fill_factor + 0.1)  # optional normalised y scale
+    plt.legend(bbox_to_anchor=(bbox, 1), loc='upper left', borderaxespad=0.)
```

Comment:
There seems to be a mix of pyplot and non-pyplot usage.  It might be preferable to stick with one approach.

Reply:
OK would need guidance as to what you suggest